gdm: refactor to properly handle wayland sessions

Elyhaka opened the pull request:
###### Motivation for this change

This PR tries to address issue #76169.

It is a draft PR since I'm currently stuck on using `systemd.packages` as mentioned by @hedning [here](https://github.com/NixOS/nixpkgs/issues/76169#issuecomment-569458028).

Here is the current state of this PR :

 - **What works** :

   - Listing GNOME session only if user explicitly enables it or if users enable `desktopManager.gnome3`
   - Listing Wayland sessions

    _If you're using a wayland compositor (like Sway) and you want to use GDM, then it works correctly in the current state._

 - **What does not work** :

    Having those 2 options enabled :
    ```
     displayManager.gdm.enable = true;
     desktopManager.gnome3.enable = true;
   ```

I have the feeling it is due to the duplication of `gnome-session` (one with the exposing session file flag, the other one not) in `systemd.packages` :

```
$ nixos-option systemd.packages
Value:
[
  «derivation /nix/store/(hash of not exposed one)-gnome-session-3.34.2.drv»
  «derivation /nix/store/(hash of exposed one)-gnome-session-3.34.2.drv»
]
```

I'm opening this PR to seek help of where I should be heading to from this point. It clearly is a naive implementation, and if you believe this is not the way it should be fixed we can simply dismiss this PR :).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @lethalman @jtojnar @hedning @worldofpeace

Commit:
gdm: refactor to properly handle wayland sessions

hedning commented on pkgs/desktops/gnome-3/core/gnome-session/default.nix:
```diff
@@ -48,6 +50,14 @@ stdenv.mkDerivation rec {
+  # We hide the GNOME sessions by default since gnome-session is a dependency of
+  # GDM itself. If we do not hide them by default, it will show broken GNOME sessions
+  # when GDM is enable without proper GNOME installation.
+  + stdenv.lib.optionalString hiddenSessions ''
```
It might be better to move the desktop files to their own special output, eg. `gnome3.gnome-session.sessions`, this can then be used in `sessionPackages`. That way we won't need to duplicate the whole derivation. 

We could also simply remove `${gnome-session}/share` from the gnome-session wrapper[0](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L36). It's there so gnome-session can find its own `share/gnome-session/sessions/gnome.session`, which is now only necessary for the x11 fallback. Though I'm guessing it might be best to keep it for now as the fallback can be useful.

Elyhaka replied:
Thank you @hedning for taking time to review !

> It might be better to move the desktop files to their own special output, eg. `gnome3.gnome-session.sessions`, this can then be used in `sessionPackages`. That way we won't need to duplicate the whole derivation.

I like this idea, I'll try to find an example of another package on `nixpkgs` to learn how to make the split. (If you have an idea on which one I could take reference from, that would be great).

> We could also simply remove `${gnome-session}/share` from the gnome-session wrapper[0](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L36). It's there so gnome-session can find its own `share/gnome-session/sessions/gnome.session`, which is now only necessary for the x11 fallback. Though I'm guessing it might be best to keep it for now as the fallback can be useful.

I think this could bring another issue : Xorg support is, as of early 2020, still a necessity for some people that are stuck with non-updated software or with proprietary drivers that are not compatible with Wayland (like nvidia with `eglstreams`).

hedning replied:
I have some code laying around, something like this should do it IIRC:
```diff
modified   pkgs/desktops/gnome-3/core/gnome-session/default.nix
@@ -6,6 +6,8 @@ stdenv.mkDerivation rec {
   pname = "gnome-session";
   version = "3.34.2";
 
+  outputs = ["out" "sessions"];
+
   src = fetchurl {
     url = "mirror://gnome/sources/gnome-session/${stdenv.lib.versions.majorMinor version}/${pname}-${version}.tar.xz";
     sha256 = "1qgqp97f8k2zi2ydvhds14zsjzfj7cv521r6wx5sw0qacn0p7dwb";
@@ -39,6 +41,12 @@ stdenv.mkDerivation rec {
     patchShebangs meson_post_install.py
   '';
 
+  postInstall = ''
+    mkdir $sessions
+    moveToOutput share/wayland-sessions "$sessions"
+    moveToOutput share/xsessions "$sessions"
+  '';
+
   # `bin/gnome-session` will reset the environment when run in wayland, we
   # therefor wrap `libexec/gnome-session-binary` instead which is the actual
   # binary needing wrapping
```

> Xorg support is, as of early 2020

Right, with the X11 fallback I'm referring to the session gnome falls back to if it fails to launch through systemd (which is what happens with GDM without gnome3.enable). I see that wasn't very clear.

But yeah, I think just having another output is best, as the fallback might be useful.

hedning commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -52,6 +52,14 @@ in
         debugging messages in GDM
       '';
 
+      forceGnomeSessions = mkOption {
```
Missed this when I looked over last time.

I'm not really convinced this option is necessary tbh. We already have an option to enable gnome sessions through the gnome desktop, so this seems a bit superfluous.

Elyhaka replied:
I thought this could be a way to restore the current behavior in case somebody would need it (even if broken). Was not convinced either tbh : I removed it.

Elyhaka commented:
I haven't been able to test if it works correctly for standard Gnome setups : I have a build failure on some Gnome3 components on current master.

hedning commented:
With `services.xserver.desktopManager.gnome3.enable = true` and `gdm` we'll get a build error around here: https://github.com/NixOS/nixpkgs/blob/09e3b355a9fe4a3cf9d170651edb8bec2ac96e6f/nixos/modules/system/boot/systemd-lib.nix#L150

The solution with the least resistance is probably adding an option to the gnome desktop that adds `gnome-session` as a systemd package, adds `gnome-shell` as a portal and turns portals on (which also adds the systemd stuff) and installs `adwaita-icon-theme`. Not very pretty though, but works around the limitation of not being able to add duplicate systemd packages.

worldofpeace commented:
> With `services.xserver.desktopManager.gnome3.enable = true` and `gdm` we'll get a build error around here:
> 
> https://github.com/NixOS/nixpkgs/blob/09e3b355a9fe4a3cf9d170651edb8bec2ac96e6f/nixos/modules/system/boot/systemd-lib.nix#L150
> 
> The solution with the least resistance is probably adding an option to the gnome desktop that adds `gnome-session` as a systemd package, adds `gnome-shell` as a portal and turns portals on (which also adds the systemd stuff) and installs `adwaita-icon-theme`. Not very pretty though, but works around the limitation of not being able to add duplicate systemd packages.

I'd much rather fix the bug of duplicates not working in `systemd.packages`.
I've ran into it before.

worldofpeace commented:
Actually, core-shell doesn't add `gnome-shell` in `systemd.packages` https://github.com/NixOS/nixpkgs/blob/9e323fbc2a5e554bce2112ed8cc0c39593f51eed/nixos/modules/services/x11/desktop-managers/gnome3.nix#L239. It being added because of being in `xdg.portals` is totally an accident and we should fix that.

worldofpeace commented:
I think I fixed the duplicates problem btw https://github.com/NixOS/nixpkgs/pull/77294.

hedning commented:
>  It's being added because of being in xdg.portals is totally an accident and we should fix that.

Yeah, with the duplicates being fixed we should add gnome-shell explicitly in core-shell. IIRC I tried adding gnome-shell to systemd.packages while we added systemd support in 3.34 and ran into the problem that gnome-shell was already added through the portal.

Elyhaka commented:
Hello, sorry, I'm a bit short at the moment : I will wait for #77294 to be merged then I will rebase this pull request. I will have more time next week to finish the work on this :).

hedning commented:
Yep, makes sense to wait for #77294, this PR should be fine then :)

worldofpeace commented:
I've merged https://github.com/NixOS/nixpkgs/pull/77294 :sparkles:

lovesegfault commented:
:tada:

Elyhaka commented:
I've rebased the branch, I'll make some tests throughout the week-end.

hedning commented:
@GrahamcOfBorg test gnome3

hedning reviewed (approved):
LGTM at least :)

worldofpeace commented:
> I've rebased the branch, I'll make some tests throughout the week-end.

Oof, gdm tests are pretty out there. I've had that on my todo list for a while.
It would be a great idea to add them, but I don't think it should block this contribution right away.
I don't believe them to be as straightforward.

worldofpeace committed:
nixos/gnome3: add gnome-shell to systemd.packages

This actually gets added via xdg.portal having gnome-shell.
However, that could change and it still for sure should be here.

worldofpeace reviewed (approved):
I started sway in a vm with
```
services.xserver.enable = true;
  services.xserver.displayManager.gdm.enable = true;
  programs.sway.enable = true;
  environment.systemPackages = [ pkgs.dmenu ];
  programs.gnome-terminal.enable = true;
  programs.dconf.enable = true;
```

from gdm.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/january-2020-in-nixos/5771/1
